### DISCUSSION TITLE
Are :matches-path() and ::after supposed to work together ?

### DISCUSSION THREAD

**Discussion by Procyon-b** [uBlockOrigin/uBlock-issues#3424] [General] [answered]
I don't want to file a bug for this. Only a FYI if the limitation isn't intended.

I've solved an [issue](https://new.reddit.com/r/uBlockOrigin/comments/1g8da8x/in_rcfb_if_you_mouse_over_the_downvote_button_on/) where, on a [subreddit](https://old.reddit.com/r/CFB/) on `old`, the hover effect on the post downvote icon displays the ::after pseudo-element.

The filter `old.reddit.com,www.reddit.com##div.arrow.down:hover::after` works as expected.

I also wanted to limit the filter to that specific sub-reddit :
`old.reddit.com,www.reddit.com##:matches-path(/r/CFB) div.arrow.down:hover::after`
but apparently `:matches-path()` and `::after` aren't allowed together.

Edit: uBO v1.60.1b15

**Comment by gorhill** [comment, marked as answer]
CSS pseudo elements can't be used with procedural cosmetic filters because `querySelectorAll()` doesn't work with these. [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):

> [CSS pseudo-elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements) will never return any elements, as specified in the [Selectors API](https://www.w3.org/TR/selectors-api/#grammar)

**Reply by garry-ut99** [reply]
> Procyon-b : Are `:matches-path()` and `::after` supposed to work together?
> Procyon-b : but apparently `:matches-path()` and `::after` aren't allowed together.
> gorhill: CSS pseudo elements (...)

It's not only about pseudo-elements (`::after`), pseudo-classes (`:hover`) give no effect either. For example this won't work either in the first place:  `old.reddit.com,www.reddit.com##:matches-path(/r/CFB) div.arrow.down:hover` (while this works: `old.reddit.com,www.reddit.com##div.arrow.down:hover`), despite uBO parser doesn't throw it as `invalid` (I'm not saying it should throw, because I don't know, I'm just simply saying what I noticed).

**Reply by Procyon-b** [reply]
for me, both `old.reddit.com,www.reddit.com##:matches-path(/r/CFB) div.arrow.down:hover`
and `old.reddit.com,www.reddit.com##div.arrow.down:hover` work
But there is a race condition for the first. It's slower to take effect.

In fact the pure css filter is in itself an infinite loop:

- hide the element when mouse is over it
- element is hidden, mouse is not considered over it anymore, -> show the element
- the element is visible, mouse is over it -> hide the element
- etc...

browsers get out of the loop by ignoring step 3 if the mouse doesn't move.

**Reply by garry-ut99** [reply]
Maybe this one: `old.reddit.com,www.reddit.com##:matches-path(/r/CFB) div.arrow.down:hover:matches-css-after(position: fixed)`. 

Or maybe can also try to replace `:matches-css-after()` with some `:xpath(something)` where `something` is supposed to be some way to emulate css `::after` (if ever possible).

**Comment by Procyon-b** [comment]
I suspected that it was caused by the implementation of :matches-path (hence, not a bug report).

**Reply by garry-ut99** [reply]
Duplicate of:
- https://github.com/uBlockOrigin/uBlock-issues/issues/238
- https://github.com/uBlockOrigin/uBlock-issues/issues/2090
- https://github.com/uBlockOrigin/uBlock-issues/issues/2786

**Reply by Procyon-b** [reply]
In my mind I don't consider :matches-path as a procedural filter similar to the others.
I'm too stylus-inclined.  ;)

**Comment by krystian3w** [comment]
https://github.com/uBlockOrigin/uBlock-issues/issues/3235#issuecomment-2787315161 https://github.com/gorhill/uBlock/commit/8b696a691a0871da8200d1806300842d988a4326

If page is static (lack of dynamic JS changes or updating uri in JS).
